mlir/lib/Dialect/Rock/utility/fusionUtils.cpp

```diff
@@ -0,0 +1,86 @@
+//===- fusionUtils.cpp - Rock utility for fusion -----------------===//
+//
+// Part of the MLIR Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===-----------------------------------------------------===//
+
+#include "mlir/Dialect/Rock/utility/fusionUtils.h"
+#include "mlir/Analysis/BufferDependencyAnalysis.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Rock/IR/Rock.h"
+#include "mlir/Dialect/Rock/Tuning/GridwiseGemmParams.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/AnalysisManager.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Support/LogicalResult.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Casting.h"
+
+using namespace mlir;
+using namespace mlir::rock;
+
+struct TestFusibilityPass
+    : public PassWrapper<TestFusibilityPass, OperationPass<func::FuncOp>> {
+  void runOnOperation() override {
+    BufferDependencyAnalysis &analysis =
+        getAnalysis<BufferDependencyAnalysis>();
+    if (failed(analysis.run())) {
+      return;
+    }
+
+    const auto &readersTable = analysis.getReadersTable();
+    const auto &writersTable = analysis.getWritersTable();
+
+    func::FuncOp func = getOperation();
+    WalkResult walkResult =
+        func.walk([&](rock::RockGemmWrapperInterface gemmOp) -> WalkResult {
+          auto gemmResult = gemmOp->getOperand(2);
+          auto allocOp = BufferDependencyAnalysis::getAllocation(gemmResult);
+          if (!allocOp.has_value()) {
+            return WalkResult::advance();
+          }
+
+          // make sure that no `linalg::GenericOp` reads from a gemm output
+          if (readersTable.contains(allocOp.value())) {
+            for (Operation *op : readersTable.at(allocOp.value())) {
```

Comment:
Since you checks whether op is contained in the vector, vector is not the right choice of data structure.

Reply:
No, I don't test whether an op in the vector. I test whether at least one reader of a given `allocOp` is `linalg::GenericOp`